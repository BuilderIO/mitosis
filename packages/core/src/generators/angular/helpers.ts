import { type MitosisComponent } from '@/types/mitosis-component';

export const HELPER_FUNCTIONS = (
  isTs?: boolean,
): {
  [key: string]: string;
} => ({
  useObjectWrapper: `useObjectWrapper(...args${isTs ? ': any[]' : ''}) {
    let obj = {}
    args.forEach((arg) => {
      obj = { ...obj, ...arg };
    });
    return obj;
  }`,
  useObjectDotValues: `useObjectDotValues(obj${isTs ? ': any' : ''})${isTs ? ': any[]' : ''}) {
    return Object.values(obj);
  }`,
  useTypeOf: `useTypeOf(obj${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return typeof obj;
  }`,
  useJsonStringify: `useJsonStringify(...args${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return JSON.stringify(...args);
  }`,
  setAttributes: `setAttributes(el${isTs ? ': HTMLElement' : ''}, value${
    isTs ? ': any' : ''
  }, changes${isTs ? '?: any' : ''}) {
    if (!el) {
      return;
    }
    const target = typeof changes === 'undefined' ? value : changes;
    Object.keys(target).forEach((key) => {
      if (key.startsWith('on')) {
        if (this._listenerFns.has(key)) {
          this._listenerFns.get(key)${isTs ? '!' : ''}();
        }
        this._listenerFns.set(key, this.renderer.listen(
          el,
          key.replace('on', '').toLowerCase(),
          target[key]
        ));
      } else {
        this.renderer.setAttribute(el, key.toLowerCase(), target[key] ?? '');
      }
    });
  }`,
});

export const getAppropriateTemplateFunctionKeys = (code: string) =>
  Object.keys(HELPER_FUNCTIONS()).filter((key) => code.includes(key));

/**
 * Adds code to the `onUpdate` hook of a MitosisComponent.
 *
 * @param {MitosisComponent} root - The root MitosisComponent.
 * @param {string} code - The code to be added to the `onUpdate` hook.
 */
export const addCodeToOnUpdate = (root: MitosisComponent, code: string) => {
  root.hooks.onUpdate = root.hooks.onUpdate || [];
  root.hooks.onUpdate.push({
    code,
  });
};

/**
 * Adds code to the `onInit` hook of a MitosisComponent.
 *
 * @param {MitosisComponent} root - The root MitosisComponent.
 * @param {string} code - The code to be added to the `onInit` hook.
 */
export const addCodeToOnInit = (root: MitosisComponent, code: string) => {
  if (!root.hooks.onInit?.code) {
    root.hooks.onInit = { code: '' };
  }
  root.hooks.onInit.code += `\n${code};`;
};

/**
 * Creates a reactive state in Angular.
 * Initializes the state with `null` because we cannot access `state.` or `props.` properties before the component is initialized.
 * Adds the code (init/re-init code) to the `onInit` and `onUpdate` hooks.
 * @param root The root MitosisComponent.
 * @param stateName The name of the reactive state.
 * @param code The code to be added to the onInit and onUpdate hooks.
 */
export const makeReactiveState = (root: MitosisComponent, stateName: string, code: string) => {
  root.state[stateName] = { code: 'null', type: 'property' };
  addCodeToOnInit(root, code);
  addCodeToOnUpdate(root, code);
};
