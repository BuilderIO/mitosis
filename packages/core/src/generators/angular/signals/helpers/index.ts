export const getAngularCoreImportsAsString = ({
  refs,
  output,
  input,
  model,
  onPush,
  effect,
  signal,
}: {
  refs: boolean;
  output: boolean;
  input: boolean;
  model: boolean;
  onPush: boolean;
  effect: boolean;
  signal: boolean;
}): string => {
  const angularCoreImports: Record<string, boolean> = {
    Component: true,
    AfterViewInit: true,
    viewChild: refs,
    ElementRef: refs,
    model,
    output,
    input,
    effect,
    signal,
    ChangeDetectionStrategy: onPush,
  };
  return Object.entries(angularCoreImports)
    .map(([key, bool]) => (bool ? key : ''))
    .filter((key) => !!key)
    .join(', ');
};
