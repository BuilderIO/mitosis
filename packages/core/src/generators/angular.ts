import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { getProps } from '../helpers/get-props';
import { kebabCase } from 'lodash';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { indent } from '../helpers/indent';

export interface ToAngularOptions extends BaseTranspilerOptions {}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAngularOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToAngular(item, options))
      .join('\n')}</div>`;
  },
};

export const blockToAngular = (
  json: MitosisNode,
  options: ToAngularOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<ng-content></ng-content>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `<ng-container *ngFor="let ${
      json.properties._forName
    } of ${stripStateAndPropsRefs(json.bindings.each as string)}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else if (json.name === 'Show') {
    str += `<ng-container *ngIf="${stripStateAndPropsRefs(
      json.bindings.when as string,
    )}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else {
    str += `<${json.name} `;

    // TODO: spread support for angular
    // if (json.bindings._spread) {
    //   str += `v-bind="${stripStateAndPropsRefs(
    //     json.bindings._spread as string,
    //   )}"`;
    // }

    for (const key in json.properties) {
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === '_spread') {
        continue;
      }
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (
          event === 'change' &&
          json.name === 'input' /* todo: other tags */
        ) {
          event = 'input';
        }
        // TODO: proper babel transform to replace. Util for this
        const finalValue = removeSurroundingBlock(
          useValue.replace(/event\./g, '$event.'),
        );
        str += ` (${event})="${finalValue}" `;
      } else if (key === 'ref') {
        str += ` #${useValue} `;
      } else {
        str += ` [${key}]="${useValue}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToAngular(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

export const componentToAngular =
  (options: ToAngularOptions = {}): Transpiler =>
  ({ component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }

    const props = getProps(component);

    const refs = Array.from(getRefs(json));
    mapRefs(json, (refName) => `this.${refName}.nativeElement`);

    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    let template = json.children.map((item) => blockToAngular(item)).join('\n');
    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      valueMapper: (code) =>
        stripStateAndPropsRefs(code, { replaceWith: 'this.' }),
    });

    let str = dedent`
    import { Component ${refs.length ? ', ViewChild, ElementRef' : ''}${
      props.size ? ', Input' : ''
    } } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      selector: '${kebabCase(json.name || 'my-component')}',
      template: \`
        ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
      \`,
      ${
        css.length
          ? `styles: [
        \`${indent(css, 8)}\`
      ],`
          : ''
      }
    })
    export default class ${component.name} {
      ${Array.from(props)
        .map((item) => `@Input() ${item}: any`)
        .join('\n')}

      ${refs
        .map((refName) => `@ViewChild('${refName}') ${refName}: ElementRef`)
        .join('\n')}

      ${
        !component.hooks.onMount
          ? ''
          : `ngOnInit() {
              ${stripStateAndPropsRefs(component.hooks.onMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${
        !component.hooks.onUpdate?.length
          ? ''
          : `ngAfterContentChecked() {
              ${component.hooks.onUpdate.map((hook) =>
                stripStateAndPropsRefs(hook.code, {
                  replaceWith: 'this.',
                }),
              )}
            }`
      }

      ${
        !component.hooks.onUnMount
          ? ''
          : `ngOnDestroy() {
              ${stripStateAndPropsRefs(component.hooks.onUnMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${dataString}
    }
  `;

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };

const tryFormat = (str: string, parser: string) => {
  try {
    return format(str, {
      parser,
      plugins: [
        // To support running in browsers
        require('prettier/parser-typescript'),
        require('prettier/parser-postcss'),
        require('prettier/parser-html'),
        require('prettier/parser-babel'),
      ],
      htmlWhitespaceSensitivity: 'ignore',
    });
  } catch (err) {
    console.warn('Could not prettify', { string: str }, err);
  }
  return str;
};
