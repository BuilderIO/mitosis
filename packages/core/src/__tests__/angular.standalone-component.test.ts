import { parseJsx } from '..';
import { componentToAngular } from '../generators/angular';
import { tryPrettierFormat } from 'src/helpers/try-prettier-format';
import basicExample from './data/basic.raw.tsx?raw';
import renderBlockExample from './data/blocks/render-block.raw.tsx?raw';
import subComponentExample from './data/sub-component.raw.tsx?raw';

const generateNgModule = (name: string, componentsUsed: string[]) =>
  tryPrettierFormat(
    `@NgModule({
  declarations: [${name}],
  imports: [CommonModule${
    componentsUsed.length ? ', ' + componentsUsed.map((comp) => `${comp}Module`).join(', ') : ''
  }],
  exports: [${name}],
})
export class ${name}Module {}`,
    'typescript',
  ).trim();

const generateStandaloneComponentMetadata = (name: string, componentsUsed: string[]) =>
  tryPrettierFormat(
    `
  standalone: true,
  imports: [CommonModule${
    componentsUsed.length ? ', ' + componentsUsed.map((comp) => `${comp}`).join(', ') : ''
  }],
})
export class ${name} {`,
    'typescript',
  ).trim();

describe('Angular standalone component', () => {
  test('Standalone option should prevent NgModule generation in basic components', () => {
    const basicNgModule = generateNgModule('MyBasicComponent', []);
    const standaloneComponentMetadata = generateStandaloneComponentMetadata('MyBasicComponent', []);
    const component = parseJsx(basicExample);

    const nonStandaloneComponent = componentToAngular({ standalone: false })({ component });
    expect(nonStandaloneComponent).toMatchSnapshot();
    expect(nonStandaloneComponent).toContain(basicNgModule);

    const standaloneComponent = componentToAngular({ standalone: true })({ component });
    expect(standaloneComponent).toMatchSnapshot();
    expect(standaloneComponent).not.toContain(basicNgModule);
    expect(standaloneComponent).toContain(standaloneComponentMetadata);
  });

  test('Standalone option should prevent NgModule generation when used whith sub-components', () => {
    const subComponentNgModule = generateNgModule('SubComponent', ['Foo']);
    const standaloneComponentMetadata = generateStandaloneComponentMetadata('SubComponent', [
      'Foo',
    ]);
    const component = parseJsx(subComponentExample);

    const nonStandaloneComponent = componentToAngular({ standalone: false })({ component });
    expect(nonStandaloneComponent).toMatchSnapshot();
    expect(nonStandaloneComponent).toContain(subComponentNgModule);

    const standaloneComponent = componentToAngular({ standalone: true })({ component });
    expect(standaloneComponent).toMatchSnapshot();
    expect(standaloneComponent).not.toContain(subComponentNgModule);
    expect(standaloneComponent).toContain(standaloneComponentMetadata);
  });

  test('Standalone option should prevent NgModule generation in complex components', () => {
    const renderBlockNgModule = generateNgModule('RenderBlock', [
      'RenderRepeatedBlock',
      'RenderBlock',
      'BlockStyles',
    ]);
    const standaloneComponentMetadata = generateStandaloneComponentMetadata('RenderBlock', [
      'RenderRepeatedBlock',
      'RenderBlock',
      'BlockStyles',
    ]);
    const component = parseJsx(renderBlockExample);

    const nonStandaloneComponent = componentToAngular({ standalone: false })({ component });
    expect(nonStandaloneComponent).toMatchSnapshot();
    expect(nonStandaloneComponent).toContain(renderBlockNgModule);

    const standaloneComponent = componentToAngular({ standalone: true })({ component });
    expect(standaloneComponent).toMatchSnapshot();
    expect(standaloneComponent).not.toContain(renderBlockNgModule);
    expect(standaloneComponent).toContain(standaloneComponentMetadata);
  });
});
