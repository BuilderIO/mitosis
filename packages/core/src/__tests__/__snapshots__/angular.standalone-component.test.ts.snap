// Vitest Snapshot v1

exports[`Angular standalone component > Standalone option should prevent NgModule generation in basic components 1`] = `
"import { NgModule } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import { Component } from \\"@angular/core\\";

export const DEFAULT_VALUES = {
  name: \\"Steve\\",
};

@Component({
  selector: \\"my-basic-component, MyBasicComponent\\",
  template: \`
    <div class=\\"test div\\">
      <input
        [attr.value]=\\"DEFAULT_VALUES.name || name\\"
        (input)=\\"name = $event.target.value\\"
      />

      Hello! I can run in React, Vue, Solid, or Liquid!
    </div>
  \`,
  styles: [
    \`
      .div {
        padding: 10px;
      }
    \`,
  ],
})
export class MyBasicComponent {
  DEFAULT_VALUES = DEFAULT_VALUES;

  name = \\"Steve\\";
  underscore_fn_name() {
    return \\"bar\\";
  }
  age = 1;
  sports = [\\"\\"];
}

@NgModule({
  declarations: [MyBasicComponent],
  imports: [CommonModule],
  exports: [MyBasicComponent],
})
export class MyBasicComponentModule {}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation in basic components 2`] = `
"import { Component } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

export const DEFAULT_VALUES = {
  name: \\"Steve\\",
};

@Component({
  selector: \\"my-basic-component, MyBasicComponent\\",
  template: \`
    <div class=\\"test div\\">
      <input
        [attr.value]=\\"DEFAULT_VALUES.name || name\\"
        (input)=\\"name = $event.target.value\\"
      />

      Hello! I can run in React, Vue, Solid, or Liquid!
    </div>
  \`,
  styles: [
    \`
      .div {
        padding: 10px;
      }
    \`,
  ],
  standalone: true,
  imports: [CommonModule],
})
export class MyBasicComponent {
  DEFAULT_VALUES = DEFAULT_VALUES;

  name = \\"Steve\\";
  underscore_fn_name() {
    return \\"bar\\";
  }
  age = 1;
  sports = [\\"\\"];
}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation in complex components 1`] = `
"import { NgModule } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import { Component, Input } from \\"@angular/core\\";

import { getBlockActions } from \\"../../functions/get-block-actions.js\\";
import { getBlockComponentOptions } from \\"../../functions/get-block-component-options.js\\";
import { getBlockProperties } from \\"../../functions/get-block-properties.js\\";
import { getBlockTag } from \\"../../functions/get-block-tag.js\\";
import { getProcessedBlock } from \\"../../functions/get-processed-block.js\\";
import { evaluate } from \\"../../functions/evaluate.js\\";
import { isEmptyHtmlElement } from \\"./render-block.helpers.js\\";
import { TARGET } from \\"../../constants/target.js\\";
import { extractTextStyles } from \\"../../functions/extract-text-styles.js\\";
import { getReactNativeBlockStyles } from \\"../../functions/get-react-native-block-styles.js\\";

@Component({
  selector: \\"render-block, RenderBlock\\",
  template: \`
    <ng-container *ngIf=\\"shouldWrap\\">
      <ng-container *ngIf=\\"isEmptyHtmlElement(tag)\\">
        <state.tag></state.tag>
      </ng-container>

      <ng-container *ngIf=\\"!isEmptyHtmlElement(tag) && repeatItemData\\">
        <ng-container *ngFor=\\"let data of repeatItemData; let index = index\\">
          <render-repeated-block
            [key]=\\"index\\"
            [repeatContext]=\\"data.context\\"
            [block]=\\"data.block\\"
          ></render-repeated-block>
        </ng-container>
      </ng-container>

      <ng-container *ngIf=\\"!isEmptyHtmlElement(tag) && !repeatItemData\\">
        <state.tag>
          <state.renderComponentTag></state.renderComponentTag>

          <ng-container *ngFor=\\"let child of childrenWithoutParentComponent\\">
            <RenderBlock
              [key]=\\"'render-block-' + child.id\\"
              [block]=\\"child\\"
              [context]=\\"childrenContext\\"
            ></RenderBlock>
          </ng-container>

          <ng-container *ngFor=\\"let child of childrenWithoutParentComponent\\">
            <block-styles
              [key]=\\"'block-style-' + child.id\\"
              [block]=\\"child\\"
              [context]=\\"childrenContext\\"
            ></block-styles>
          </ng-container>
        </state.tag>
      </ng-container>
    </ng-container>
  \`,
})
export class RenderBlock {
  isEmptyHtmlElement = isEmptyHtmlElement;

  @Input() block: any;
  @Input() context: any;

  get component() {
    const componentName = getProcessedBlock({
      block: this.block,
      this: this.context.state,
      context: this.context.context,
      shouldEvaluateBindings: false,
    }).component?.name;

    if (!componentName) {
      return null;
    }

    const ref = this.context.registeredComponents[componentName];

    if (!ref) {
      // TODO: Public doc page with more info about this message
      console.warn(\`
          Could not find a registered this.component named \\"\${componentName}\\".
          If you registered it, is the file that registered it imported by the file that needs to render it?\`);
      return undefined;
    } else {
      return ref;
    }
  }
  get tag() {
    return getBlockTag(this.useBlock);
  }
  get useBlock() {
    return this.repeatItemData
      ? this.block
      : getProcessedBlock({
          block: this.block,
          this: this.context.state,
          context: this.context.context,
          shouldEvaluateBindings: true,
        });
  }
  get actions() {
    return getBlockActions({
      block: this.useBlock,
      this: this.context.state,
      context: this.context.context,
    });
  }
  get attributes() {
    const blockProperties = getBlockProperties(this.useBlock);
    return {
      ...blockProperties,
      ...(TARGET === \\"reactNative\\"
        ? {
            style: getReactNativeBlockStyles({
              block: this.useBlock,
              context: this.context,
              blockStyles: blockProperties.style,
            }),
          }
        : {}),
    };
  }
  get shouldWrap() {
    return !this.component?.noWrap;
  }
  get renderComponentProps() {
    return {
      blockChildren: this.useChildren,
      componentRef: this.component?.component,
      componentOptions: {
        ...getBlockComponentOptions(this.useBlock),

        /**
         * These this.attributes are passed to the wrapper element when there is one. If \`noWrap\` is set to true, then
         * they are provided to the this.component itself directly.
         */
        ...(this.shouldWrap
          ? {}
          : {
              attributes: { ...this.attributes, ...this.actions },
            }),
        customBreakpoints: this.childrenContext?.content?.meta?.breakpoints,
      },
      context: this.childrenContext,
    };
  }
  get useChildren() {
    // TO-DO: When should \`canHaveChildren\` dictate rendering?
    // This is currently commented out because some Builder components (e.g. Box) do not have \`canHaveChildren: true\`,
    // but still receive and need to render children.
    // return state.componentInfo?.canHaveChildren ? state.useBlock.children : [];
    return this.useBlock.children ?? [];
  }
  get childrenWithoutParentComponent() {
    /**
     * When there is no \`componentRef\`, there might still be children that need to be rendered. In this case,
     * we render them outside of \`componentRef\`.
     * NOTE: We make sure not to render this if \`repeatItemData\` is non-null, because that means we are rendering an array of
     * blocks, and the children will be repeated within those blocks.
     */
    const shouldRenderChildrenOutsideRef =
      !this.component?.component && !this.repeatItemData;
    return shouldRenderChildrenOutsideRef ? this.useChildren : [];
  }
  get repeatItemData() {
    /**
     * we don't use \`state.useBlock\` here because the processing done within its logic includes evaluating the block's bindings,
     * which will not work if there is a repeat.
     */
    const { repeat, ...blockWithoutRepeat } = this.block;

    if (!repeat?.collection) {
      return undefined;
    }

    const itemsArray = evaluate({
      code: repeat.collection,
      this: this.context.state,
      context: this.context.context,
    });

    if (!Array.isArray(itemsArray)) {
      return undefined;
    }

    const collectionName = repeat.collection.split(\\".\\").pop();
    const itemNameToUse =
      repeat.itemName || (collectionName ? collectionName + \\"Item\\" : \\"item\\");
    const repeatArray = itemsArray.map((item, index) => ({
      context: {
        ...this.context,
        this: {
          ...this.context.state,
          $index: index,
          $item: item,
          [itemNameToUse]: item,
          [\`$\${itemNameToUse}Index\`]: index,
        },
      },
      block: blockWithoutRepeat,
    }));
    return repeatArray;
  }
  get inheritedTextStyles() {
    if (TARGET !== \\"reactNative\\") {
      return {};
    }

    const styles = getReactNativeBlockStyles({
      block: this.useBlock,
      context: this.context,
      blockStyles: this.attributes.style,
    });
    return extractTextStyles(styles);
  }
  get childrenContext() {
    return {
      apiKey: this.context.apiKey,
      this: this.context.state,
      content: this.context.content,
      context: this.context.context,
      registeredComponents: this.context.registeredComponents,
      inheritedStyles: this.inheritedTextStyles,
    };
  }
  get renderComponentTag() {
    if (TARGET === \\"reactNative\\") {
      return RenderComponentWithContext;
    } else if (TARGET === \\"vue3\\") {
      // vue3 expects a string for the this.component tag
      return \\"RenderComponent\\";
    } else {
      return RenderComponent;
    }
  }
}

@NgModule({
  declarations: [RenderBlock],
  imports: [
    CommonModule,
    RenderRepeatedBlockModule,
    RenderBlockModule,
    BlockStylesModule,
  ],
  exports: [RenderBlock],
})
export class RenderBlockModule {}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation in complex components 2`] = `
"import { Component, Input } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import { getBlockActions } from \\"../../functions/get-block-actions.js\\";
import { getBlockComponentOptions } from \\"../../functions/get-block-component-options.js\\";
import { getBlockProperties } from \\"../../functions/get-block-properties.js\\";
import { getBlockTag } from \\"../../functions/get-block-tag.js\\";
import { getProcessedBlock } from \\"../../functions/get-processed-block.js\\";
import { evaluate } from \\"../../functions/evaluate.js\\";
import BlockStyles from \\"./block-styles\\";
import { isEmptyHtmlElement } from \\"./render-block.helpers.js\\";
import RenderRepeatedBlock from \\"./render-repeated-block\\";
import { TARGET } from \\"../../constants/target.js\\";
import { extractTextStyles } from \\"../../functions/extract-text-styles.js\\";
import RenderComponentWithContext from \\"./render-component-with-context\\";
import RenderComponent from \\"./render-component\\";
import { getReactNativeBlockStyles } from \\"../../functions/get-react-native-block-styles.js\\";

@Component({
  selector: \\"render-block, RenderBlock\\",
  template: \`
    <ng-container *ngIf=\\"shouldWrap\\">
      <ng-container *ngIf=\\"isEmptyHtmlElement(tag)\\">
        <state.tag></state.tag>
      </ng-container>

      <ng-container *ngIf=\\"!isEmptyHtmlElement(tag) && repeatItemData\\">
        <ng-container *ngFor=\\"let data of repeatItemData; let index = index\\">
          <render-repeated-block
            [key]=\\"index\\"
            [repeatContext]=\\"data.context\\"
            [block]=\\"data.block\\"
          ></render-repeated-block>
        </ng-container>
      </ng-container>

      <ng-container *ngIf=\\"!isEmptyHtmlElement(tag) && !repeatItemData\\">
        <state.tag>
          <state.renderComponentTag></state.renderComponentTag>

          <ng-container *ngFor=\\"let child of childrenWithoutParentComponent\\">
            <RenderBlock
              [key]=\\"'render-block-' + child.id\\"
              [block]=\\"child\\"
              [context]=\\"childrenContext\\"
            ></RenderBlock>
          </ng-container>

          <ng-container *ngFor=\\"let child of childrenWithoutParentComponent\\">
            <block-styles
              [key]=\\"'block-style-' + child.id\\"
              [block]=\\"child\\"
              [context]=\\"childrenContext\\"
            ></block-styles>
          </ng-container>
        </state.tag>
      </ng-container>
    </ng-container>
  \`,
  standalone: true,
  imports: [CommonModule, RenderRepeatedBlock, RenderBlock, BlockStyles],
})
export class RenderBlock {
  isEmptyHtmlElement = isEmptyHtmlElement;

  @Input() block: any;
  @Input() context: any;

  get component() {
    const componentName = getProcessedBlock({
      block: this.block,
      this: this.context.state,
      context: this.context.context,
      shouldEvaluateBindings: false,
    }).component?.name;

    if (!componentName) {
      return null;
    }

    const ref = this.context.registeredComponents[componentName];

    if (!ref) {
      // TODO: Public doc page with more info about this message
      console.warn(\`
          Could not find a registered this.component named \\"\${componentName}\\".
          If you registered it, is the file that registered it imported by the file that needs to render it?\`);
      return undefined;
    } else {
      return ref;
    }
  }
  get tag() {
    return getBlockTag(this.useBlock);
  }
  get useBlock() {
    return this.repeatItemData
      ? this.block
      : getProcessedBlock({
          block: this.block,
          this: this.context.state,
          context: this.context.context,
          shouldEvaluateBindings: true,
        });
  }
  get actions() {
    return getBlockActions({
      block: this.useBlock,
      this: this.context.state,
      context: this.context.context,
    });
  }
  get attributes() {
    const blockProperties = getBlockProperties(this.useBlock);
    return {
      ...blockProperties,
      ...(TARGET === \\"reactNative\\"
        ? {
            style: getReactNativeBlockStyles({
              block: this.useBlock,
              context: this.context,
              blockStyles: blockProperties.style,
            }),
          }
        : {}),
    };
  }
  get shouldWrap() {
    return !this.component?.noWrap;
  }
  get renderComponentProps() {
    return {
      blockChildren: this.useChildren,
      componentRef: this.component?.component,
      componentOptions: {
        ...getBlockComponentOptions(this.useBlock),

        /**
         * These this.attributes are passed to the wrapper element when there is one. If \`noWrap\` is set to true, then
         * they are provided to the this.component itself directly.
         */
        ...(this.shouldWrap
          ? {}
          : {
              attributes: { ...this.attributes, ...this.actions },
            }),
        customBreakpoints: this.childrenContext?.content?.meta?.breakpoints,
      },
      context: this.childrenContext,
    };
  }
  get useChildren() {
    // TO-DO: When should \`canHaveChildren\` dictate rendering?
    // This is currently commented out because some Builder components (e.g. Box) do not have \`canHaveChildren: true\`,
    // but still receive and need to render children.
    // return state.componentInfo?.canHaveChildren ? state.useBlock.children : [];
    return this.useBlock.children ?? [];
  }
  get childrenWithoutParentComponent() {
    /**
     * When there is no \`componentRef\`, there might still be children that need to be rendered. In this case,
     * we render them outside of \`componentRef\`.
     * NOTE: We make sure not to render this if \`repeatItemData\` is non-null, because that means we are rendering an array of
     * blocks, and the children will be repeated within those blocks.
     */
    const shouldRenderChildrenOutsideRef =
      !this.component?.component && !this.repeatItemData;
    return shouldRenderChildrenOutsideRef ? this.useChildren : [];
  }
  get repeatItemData() {
    /**
     * we don't use \`state.useBlock\` here because the processing done within its logic includes evaluating the block's bindings,
     * which will not work if there is a repeat.
     */
    const { repeat, ...blockWithoutRepeat } = this.block;

    if (!repeat?.collection) {
      return undefined;
    }

    const itemsArray = evaluate({
      code: repeat.collection,
      this: this.context.state,
      context: this.context.context,
    });

    if (!Array.isArray(itemsArray)) {
      return undefined;
    }

    const collectionName = repeat.collection.split(\\".\\").pop();
    const itemNameToUse =
      repeat.itemName || (collectionName ? collectionName + \\"Item\\" : \\"item\\");
    const repeatArray = itemsArray.map((item, index) => ({
      context: {
        ...this.context,
        this: {
          ...this.context.state,
          $index: index,
          $item: item,
          [itemNameToUse]: item,
          [\`$\${itemNameToUse}Index\`]: index,
        },
      },
      block: blockWithoutRepeat,
    }));
    return repeatArray;
  }
  get inheritedTextStyles() {
    if (TARGET !== \\"reactNative\\") {
      return {};
    }

    const styles = getReactNativeBlockStyles({
      block: this.useBlock,
      context: this.context,
      blockStyles: this.attributes.style,
    });
    return extractTextStyles(styles);
  }
  get childrenContext() {
    return {
      apiKey: this.context.apiKey,
      this: this.context.state,
      content: this.context.content,
      context: this.context.context,
      registeredComponents: this.context.registeredComponents,
      inheritedStyles: this.inheritedTextStyles,
    };
  }
  get renderComponentTag() {
    if (TARGET === \\"reactNative\\") {
      return RenderComponentWithContext;
    } else if (TARGET === \\"vue3\\") {
      // vue3 expects a string for the this.component tag
      return \\"RenderComponent\\";
    } else {
      return RenderComponent;
    }
  }
}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation in components with sub-components 1`] = `
"import { NgModule } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"sub-component, SubComponent\\",
  template: \`
    <foo></foo>
  \`,
})
export class SubComponent {}

@NgModule({
  declarations: [SubComponent],
  imports: [CommonModule, FooModule],
  exports: [SubComponent],
})
export class SubComponentModule {}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation when used whith sub-components 1`] = `
"import { NgModule } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"sub-component, SubComponent\\",
  template: \`
    <foo></foo>
  \`,
})
export class SubComponent {}

@NgModule({
  declarations: [SubComponent],
  imports: [CommonModule, FooModule],
  exports: [SubComponent],
})
export class SubComponentModule {}
"
`;

exports[`Angular standalone component > Standalone option should prevent NgModule generation when used whith sub-components 2`] = `
"import { Component } from \\"@angular/core\\";
import { CommonModule } from \\"@angular/common\\";

import Foo from \\"./foo-sub-component\\";

@Component({
  selector: \\"sub-component, SubComponent\\",
  template: \`
    <foo></foo>
  \`,
  standalone: true,
  imports: [CommonModule, Foo],
})
export class SubComponent {}
"
`;
